sqld/src/replication/http.rs

```diff
@@ -0,0 +1,229 @@
+    };
+
+    let resp = Response::builder()
+        .status(hyper::StatusCode::OK)
+        .body(Body::from(serde_json::to_vec(&hello)?))
+        .unwrap();
+    Ok(resp)
+}
+
+async fn handle_query(
+    mut req: Request<Body>,
+    _auth: crate::auth::Authenticated,
+    logger: Arc<ReplicationLogger>,
+) -> Result<Response<Body>> {
+    const MAX_FRAMES_IN_SINGLE_RESPONSE: usize = 256;
+
+    let bytes = hyper::body::to_bytes(req.body_mut()).await?;
+    let FramesRequest { next_offset } = match serde_json::from_slice(&bytes) {
+        Ok(req) => req,
+        Err(resp) => return Ok(error(&resp.to_string(), hyper::StatusCode::BAD_REQUEST)),
+    };
+    tracing::trace!("Requested next offset: {next_offset}");
+
+    let next_offset = std::cmp::max(next_offset, 1); // Frames start from 1
+    let current_frameno = next_offset - 1;
+    let mut frame_stream = FrameStream::new(logger.clone(), current_frameno);
+    tracing::trace!(
+        "Max available frame_no: {}",
+        frame_stream.max_available_frame_no
+    );
+    if frame_stream.max_available_frame_no < next_offset {
+        tracing::trace!("No frames available starting {next_offset}, returning 204 No Content");
+        return Ok(Response::builder()
+            .status(hyper::StatusCode::NO_CONTENT)
+            .body(Body::empty())
+            .unwrap());
+    }
+
+    let mut frames = Frames::new();
+    for _ in 0..MAX_FRAMES_IN_SINGLE_RESPONSE {
+        use futures::StreamExt;
+
+        match frame_stream.next().await {
+            Some(Ok(frame)) => {
+                tracing::trace!("Read frame {}", frame_stream.current_frame_no);
+                frames.push(frame);
+            }
+            Some(Err(LogReadError::SnapshotRequired)) => {
+                drop(frame_stream);
+                if frames.is_empty() {
+                    tracing::debug!("Snapshot required, switching to snapshot mode");
+                    frames = load_snapshot(logger, next_offset)?;
+                } else {
+                    tracing::debug!("Snapshot required, but some frames were read - returning.");
+                }
+                break;
+            }
+            Some(Err(e)) => {
+                tracing::error!("Error reading frame: {}", e);
+                return Ok(Response::builder()
+                    .status(hyper::StatusCode::INTERNAL_SERVER_ERROR)
+                    .body(Body::empty())
+                    .unwrap());
+            }
+            None => break,
+        }
+
+        if frame_stream.max_available_frame_no <= frame_stream.current_frame_no {
+            break;
+        }
+    }
+
+    if frames.is_empty() {
+        return Ok(Response::builder()
+            .status(hyper::StatusCode::NO_CONTENT)
+            .body(Body::empty())
+            .unwrap());
+    }
+
+    Ok(Response::builder()
+        .status(hyper::StatusCode::OK)
+        .body(Body::from(serde_json::to_string(&frames)?))
+        .unwrap())
+}
```

Comment:
ah yeah, makes perfect sense, will do

Reply:
@MarinPostma I'm inclined to stall returning a stream here, because of the current convoluted logic when a snapshot is needed - then, we either return whatever frames we have so far, or just ship the whole snapshot, and that whole magic logic would need to be wrapped in a single struct that implements futures::Stream. Doable of course, but perhaps in interation 2?